README
NAME
    MarpaX::Languages::C::AST - Translate a C source to an AST

VERSION
    version 0.09

SYNOPSIS
        use strict;
        use warnings FATAL => 'all';
        use MarpaX::Languages::C::AST;
        use Log::Log4perl qw/:easy/;
        use Log::Any::Adapter;
        use Log::Any qw/$log/;
        use Data::Dumper;
        #
        # Init log
        #
        our $defaultLog4perlConf = '
        log4perl.rootLogger              = WARN, Screen
        log4perl.appender.Screen         = Log::Log4perl::Appender::Screen
        log4perl.appender.Screen.stderr  = 0
        log4perl.appender.Screen.layout  = PatternLayout
        log4perl.appender.Screen.layout.ConversionPattern = %d %-5p %6P %m{chomp}%n
        ';
        Log::Log4perl::init(\$defaultLog4perlConf);
        Log::Any::Adapter->set('Log4perl');
        #
        # Parse C
        #
        my $cSourceCode = '
        typedef struct s1_ {int i1;} x1, y1;
        struct x1 {x1 i2;};
        x1 x;
        ';
        my $cAstObject = MarpaX::Languages::C::AST->new();
        print Dumper($cAstObject->parse(\$cSourceCode));

DESCRIPTION
    This modules translates a C source into an AST tree. The AST consist of
    blessed objects that map exactly to the C grammar in use. If you want to
    enable logging, be aware that this module is a Log::Any thingy.

    Please note that this module just *translates* a C source, it does *not*
    check for its correctness, i.e. the numerous grammar constraints built
    on top on the C grammar are not implemented, for example constraint on
    the number of storage class specifiers, uniqueness of labeled statements
    within a function, etc.. This is left to a compiler, which is not the
    goal here. So, to state things clearly, this module is addressing the
    *ambiguities* of the grammar itself, i.e. the dangling else, the
    typedef/enum/identifier. And produces an AST of the parse tree value.

SUBROUTINES/METHODS
  new($class, $grammarName)
    Instantiate a new object. Takes as parameter an optional base name of a
    grammar. Default is 'ISO-ANSI-C-2011'.

  parse($self, $referenceToSourceCodep, $optionalArrayOfValuesb)
    Do the parsing and return the blessed value. Takes as first parameter
    the reference to a C source code. Takes as optional second parameter a
    flag saying if the return value should be an array of all values or not.
    If this flag is false, the module will croak if there more than one
    parse tree value.

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://rt.cpan.org/Public/Dist/Display.html?Name=MarpaX-Languages-C-AS
    T>. You will be notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/jddurand/marpax-languages-c-ast>

      git clone git://github.com/jddurand/marpax-languages-c-ast.git

AUTHOR
    Jean-Damien Durand <jeandamiendurand@free.fr>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Jean-Damien Durand.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changelog for MarpaX-Languages-C-AST

0.07 2013-06-02T17:37:25
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - oups back


v0.06 2013-06-02T17:33:00
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Closer to DAGOLDEN


v0.05 2013-06-02T17:12:10
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Removed non needed dzil.sh, this also allow to have a clean tag, but
   Dist::Zilla perturbed a bit the followup of tag v.s. official releases


v0.04 2013-06-02T17:03:14
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Uncommentedd UploadToCPAN

 - Remove dist.ini from allow_dirty, Added back commented [FakeRelease]

 - New CPAN release of MarpaX-Languages-C-AST - v0.04
   
    [Jean-Damien Durand <jeandamiendurand@free.fr>]
    - Added META.json and README.pod to .gitignore
   
    - Add .build to .gitignore
   
    - Dist:Zilla version
   
    - Removed
   
    - Removed README boilerplate
   
    - Remove given/when: warnings on perl5.18
   
    - Better readibility

 - Added META.json and README.pod to .gitignore

 - Add .build to .gitignore

 - Dist:Zilla version

 - Removed

 - Removed README boilerplate

 - Remove given/when: warnings on perl5.18

 - Better readibility


v0.03 2013-05-30T19:12:07
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - v0.03

 - Synopsis made readable

 - Removed

 - Move atomic stuff to example since the grammar does not have pb with it
   (historically c2011 draft had an ambiguity between atomic as type
   specifier and type qualifier)

 - Updated test file to follow logging that now have a context free string

 - Use given/when, log the context (lexeme, event)

 - MIN_PERL_VERSION => '5.10.0'

 - Fix under taint mode

 - Preview of changes for 0.03

 - My own version of changes.t

 - Cosmetic reorder in _doLexeme()

 - Revisited lexeme and events processing

 - Again manifest.t
   
   Revisited any-OS (well, Windows) support

 - Fixed MANIFEST

 - Back to quasi-original manifest.t

 - TODO list, currently contains only PERHAPS items

 - Revisited log mechanism and inspectG1

 - Better log

 - examples/synopsis.pl

 - Variable name convention fix

 - Correct parameters to findInProgress()

 - No (not-needed) $lhs nor [@rhs] in $matchesp

 - findInProgressShort and g1Describe helper methods

 - Return list of [$ruleId, $dotPosition, $origin, $lhs, [@rhs]] that
   matched

 - Force Marpa::R2 in addition to the test in Makefile.PL

 - Fix rc=1 if anything was requested /and/ matched

 - Typo in a comment

 - Removed perl version dependency

 - Removed duplicate BUILD_REQUIRES

 - Do a PREREQ_PM

 - Removed non-needed Module::Info requirement

 - Removed use perl 5.006

 - Added use warnings

 - Updated methods docs

 - Updated methods docs

 - Fixed typos in the doc

 - Marpa::R2 2.056000 official release dependency

 - Add ruleid and origin in findinprogress

 - Forgot 0.02 Changes update


v0.02 2013-05-21T22:39:50
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Bumped to version 0.02

 - Fixed synopsis

 - Add repository to META_MERGE

 - Say what this module is not

 - Move to Storage (core module) instead of Clone


v0.01 2013-05-21T12:19:33
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Add package tarballs to .gitignore

 - test blessed structure result

 - Oups, removed a Data::Dumper call

 - No more untaint and if-if-else ambiguity solving

 - Fix t/Grammar/synposis.t on Win32

 - Precisions in the POD for read method

 - Untaint the grammar when reading it, use core IO::File instead of
   File::Slurp

 - Simplified new()

 - Fixed synopsis

 - Renamed if-then-else.t to if-if-else.t

 - use Marpa::R2 2.055_003

 - Fix output v.s. arrayOfValuesb

 - Possiblity to return all values

 - Fix some POD sections

 - Hmm, I messed up last commit logs - let's restart from here

 - if-then-else test
   Fixed some SEE ALSO sections

 - Few fixes and global t/synopsis.t file for the whole package

 - Log::Any wrapper on top of Marpa

 - Added $grammar->list() method

 - Grammar test files and implementation on top of Marpa

 - Use $log
   Remove __DATA__

 - MarpaX::Languages::C::AST::Grammar package
   Fixes for newest Test::More

 - Scope synopsis test

 - Scope module

 - Initial commit




Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Translate a C source to an AST",
  "AUTHOR" => "Jean-Damien Durand <jeandamiendurand\@free.fr>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "MarpaX-Languages-C-AST",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MarpaX::Languages::C::AST",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Exporter" => 0,
    "Log::Any" => 0,
    "Marpa::R2" => "2.056000",
    "Storable" => 0,
    "constant" => 0,
    "diagnostics" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Find" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "List::Util" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.09",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



